File: src/cpp/word-search.cpp
class Solution {
public:
    bool help(vector<vector<char>>& board,int cl,int i,int j,int &r,int &c,int &l,string &word){
        if(cl==l){
            return true;
        }
        // if(i==r || j==c) return false;
        char z=board[i][j];
        board[i][j]='*';
        
        if(i>0){
            if(word[cl]==board[i-1][j]) {
                if(help(board,cl+1,i-1,j,r,c,l,word)) return true;
            }
        }
        if(j>0){
            if(word[cl]==board[i][j-1]) {
                if(help(board,cl+1,i,j-1,r,c,l,word)) return true;
            }
        }
        if(i+1<r){
            if(word[cl]==board[i+1][j]) {
                if(help(board,cl+1,i+1,j,r,c,l,word)) return true;
            }
        }
        if(j+1<c){
            if(word[cl]==board[i][j+1]) {
                if(help(board,cl+1,i,j+1,r,c,l,word)) return true;
            }
        }
        board[i][j]=z;
        return false;
        
    }
    bool exist(vector<vector<char>>& board, string word) {
        int r=board.size(),c=board[0].size(),l=word.size();
        if(r*c < l) return 0;
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                if(word[0]==board[i][j]){
                    if(help(board,1,i,j,r,c,l,word)) return 1;
                }
            }
        }
        return 0;
    }
};

File: src/java/word-search.java
/*Question:-Given an m x n grid of characters board and a string word, return true if word exists in the grid.

The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.

 */

//Solution

class Solution {

	boolean search(char[][] board, char[] chars, int c, int i, int j, boolean[][] visited) {

		int p = board.length;
		int q = board[0].length;

		int numChars = chars.length;

		// Index out of bounds or current char on board is not equal to char at c
		if(c > numChars - 1 || board[i][j] != chars[c]) {
			return false;
		}

		// The destination has been reached
		if(c == numChars - 1 && board[i][j] == chars[c] && !visited[i][j]) {
			return true;
		} 

		boolean res = false;

		int n = i > 0 ? i - 1 : 0;
		int w = j > 0 ? j - 1 : 0;
		int s = i < p - 1 ? i + 1 : p - 1;
		int e = j < q - 1 ? j + 1 : q - 1;

		// Search for next potential candidate location
		if(!visited[i][j]) {
			visited[i][j] = true;
			res = res || search(board, chars, c + 1, n, j, visited);
			res = res || search(board, chars, c + 1, i, w, visited);
			res = res || search(board, chars, c + 1, s, j, visited);
			res = res || search(board, chars, c + 1, i, e, visited);
			visited[i][j] = false;
		}

		return res;
	}

	public boolean exist(char[][] board, String word) {

		int p = board.length;
		int q = board[0].length;

		boolean[][] visited = new boolean[p][q];
		char[] chars = word.toCharArray();

		boolean res = false;

		// Compute if word is in board
		for(int i = 0; i < p; ++i) {
			for(int j = 0; j < q; ++j) {
				res = res || search(board, chars, 0, i, j, visited);
			}
		}

		return res;
	}
}

//Submission link:-https://leetcode.com/submissions/detail/797204452/
